### PR TITLE
[WIP] all-to-all reshuffling in DTensor redistribute

### PR DESCRIPTION
**Pull request by XilunWu**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #94235

1. add all-to-all to multi-threaded PG to enable testing
2. test padding

### CODE REVIEW

File: test/distributed/_tensor/test_redistribute.py [hunk 1]
```diff
@@ -267,6 +268,80 @@ def test_multi_dim_mesh(self):
                         expected = num_sums * full_tensor
                         self.assertEqual(local_full, expected)
 
+    @with_comms
```
Comment (XilunWu) [inline test/distributed/_tensor/test_redistribute.py:271]: This test is out of scope. needs to be removed

File: torch/distributed/_tensor/redistribute.py [hunk 2]
```diff
@@ -140,10 +139,23 @@ def _redistribute_with_local_tensor(
...
+                        contiguous=True,
                     )
+                    print(f"before exchange: {shards}")
+                    shards = target_placement._exchange_tensor(
+                        shards,
+                        num_chunks,
+                        device_mesh,
+                        i,
+                        #with_padding=False,
+                        #contiguous=False,
```
Comment (XilunWu) [inline torch/distributed/_tensor/redistribute.py:155]: `contiguous` has to be `True`

**Review by wanchaol — COMMENTED** [review]
first pass, looks like in the right path, see comments inlined.
